core/src/main/java/org/elasticsearch/action/index/IndexRequest.java

```diff
@@ -549,8 +558,17 @@ public void writeTo(StreamOutput out) throws IOException {
     public String toString() {
         String sSource = "_na_";
         try {
-            sSource = XContentHelper.convertToJson(source, false);
-        } catch (Exception e) {
+            if (source.length() > MAX_SOURCE_LENGTH_IN_TOSTRING) {
+                // We might slice in the middle of a UTF-8 character here, but BytesRef.utf8ToString() will read more bytes if needed.
+                // There is no risk of ArrayIndexOutOfBoundsException either as source is assumed to be properly UTF-8 encoded.
+                sSource = source.slice(0, MAX_SOURCE_LENGTH_IN_TOSTRING).utf8ToString() + "..." +
+                    new ByteSizeValue(source.length()).toString();
```

Comment:
I can change that, no problem.
The only concern is that we don't know the total number of 'characters' as the `source` length is in bytes. That makes the message not very accurate.

How about: `Source is too big (10122123 bytes), starts with: _slice_goes_here_...`?
Also do you think it should print the exact number of bytes or the shorter '9.7mb' will be ok?